drud/ddev
# Merge the contents of config.yaml and `config.*.yaml`, fixes #3598

## The Problem/Issue/Bug:

* #3598

Before this fix, adding a web_environment key to an override config file (`config.SOMENAME.yaml`) would completely replace the content of that string list of environment variables. A fix needs to allow some reasonable notion of a merge, preserving old keys if not restated, updating them if changed.

## How this PR Solves The Problem:

This PR modifies ddevapp.ReadConfig() to special case the web_environment item in the config data structure. In the WebEnvironment case, for `config.yaml` and `config.override.yaml `(any overriding name),

* The original keys are added to a list
* The new keys are compared with an regexp and will match the key portion of `key=value`. If the key matches for an existing item, the new key=value will replace it.
* Any item in the new version of web_environment that did not match any in the old version will be put at the end of the list.

The PR also adds a bit of testdata and a unit test that confirms
* Old items that are not matched are preserved.
* New items with the same key will replace the old.
* Other new items are indeed added.

## Manual Testing Instructions:

1. Note the contents of the web_environment section of your config.yaml file.
2. Add these test keys to that section, say, 
```yaml
web_environment:
- TESTKEY1=oldvalue
- TESTKEY2=notchanged
```
3. Create a config.some-name.yaml file with at least this structure.
```yaml
web_environment:
- TESTKEY1=newvalue
- TESTKEY3=notinold
```
4. Run `ddev restart`, and do `ddev exec printenv | grep TESTKEY` to see the keys on the web container.
5. You should see TESTKEY1 with its new value, TESTKEY2 preserved and with its original value, and TESTKEY3 as well.

## Automated Testing Overview:

The PR  adds a bit of testdata and a unit test that confirm the fix:
* Test data is in pkg/ddevapp/testdata/TestConfigMergeItems/.ddev and includes a simple config.yaml and config.override.yml file, both implementing the web_environment key.
* Unit test `TestConfigMergeItems()` that has asserts for the following:
  * Old items that are not matched are preserved.
  * New items with the same key will replace the old.
  * Other new items are indeed added.

## Related Issue Link(s):

## Release/Deployment notes:

In all cases other than web_environment, the current clobber behavior is maintained. Anyone with a config file that duplicates the keys in config.yaml should see no change in behavior, since the new code will guarantee that all the keys in the override file will still be present and as they are in the override file.

I don't see any need anything different done in deployment unless you want to override the web_environment.




<a href="https://gitpod.io/#https://github.com/drud/ddev/pull/3983"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
